server/utils/requestType.ts

```diff
@@ -0,0 +1,29 @@
+import { ServerResponseInterface } from '../graphql/serverResponse'
+import { updateRequestGroupHelper } from './requestGroup'
+
+import { UserInputError } from 'apollo-server-errors'
+
+const updateRequestTypeHelper = (requestType, dataSources, dateUpdated = Date.now()): Promise<ServerResponseInterface> => {
+  if(!requestType.id) {
+    throw new UserInputError('Missing argument value', { argumentName: 'id' })
+  }
+  const requestGroupId = dataSources.requestTypes.getById(requestType.id).requestGroup
+  requestType.dateUpdated = dateUpdated
```

Comment:
is there ever going to be a case where dateUpdated != Date.now()?

Reply:
dateUpdated != Date.now() for `updateRequestTypeHelper` when we call it from `updateRequestHelper`

Reply:
I'm not sure whether we should start caring about times instead of dates. But passing in dateUpdated can be prone to race conditions lol, for both date and time, but more importantly for time.

E.g: update request A of requestType C at 11:59:59pm and update request B of requestType C 12:00:01am, but due to network issues or Mongo's locks, request B's update is processed first. Then requestType C has dateUpdated set to 12:00:01am. After, requestA's update occurs and requestType C has dateUpdated to 11:59:59pm, which is before requestB.

By setting dateUpdated to refer to when an object itself is updated, we guarantee that the dateUpdated cannot be chronologically before any contained object.

It's just a small issue, but it's interesting to consider.

Reply:
I'm also kind of arguing that we should configure Mongo/Mongoose to always update dateUpdated to the current date through: https://docs.mongodb.com/manual/reference/operator/update/currentDate/

Reply:
Lol I literally never would have thought of that. Maybe a way to make sure that doesn't happen is if the dateUpdated argument is provided and if it's < the current dateUpdated of the requestType, then don't update requestType.dateUpdated.

Reply:
thanks @guy-ro  for being our race condition police 🚨

Right now, our implementation for syncing the `dateUpdated` of RequestGroup, RequestType, and Request might introduce race conditions. For now, the **race condition won't introduce any other bugs** because we don't have any critical logic that depends on `dateUpdated` being the most recent time the document was updated.

In the future, however, we might have critical logic that depends on `dateUpdated` being accurate. (lol i don't want anyone trying to troubleshoot race conditions)

To future-proof our code, I propose this:
+ Take out the `dateUpdated` param from all update helper methods
+ In `CachedMongooseDataSource.update`, use `findByIdAndUpdate(inputObject.id, inputObject)`
+ Use schema level timestamps for RequestGroup, RequestType, and Request. These timestamps will update automatically whenever you update the document. See docs [here](https://mongoosejs.com/docs/guide.html#timestamps) and [here](https://mongoosejs.com/docs/api.html#model_Model.findByIdAndUpdate)
```
const RequestTypeSchema = new Schema({
  //...
  dateUpdated: {
    type: Date,
    required: true,
    default: Date.now()
  },
  //...
},
{
  timestamps: { 
    currentTime: Date.now,
    updatedAt: "dateUpdated"
  }
})
```

Reply:
Note that findByIdAndUpdate doesn't return any error if it fails to find the id. So we need error checking in datasources if we use findByIdAndUpdate instead of update

Reply:
@meganniu why wouldn't using update work?

Reply:
Also, note that if dateUpdated is set on the object passed to update/findByIdAndUpdate then the timestamp may fail to update (so ensure that this property isn't set).

Reply:
> @meganniu why wouldn't using update work?

I'm assuming you're referring to Mongoose's [`Model.update`](https://mongoosejs.com/docs/api.html#model_Model.update). Currently, `CachedMongooseDataSource.update` uses `Model.findByIdAndUpdate`. 

I'm assuming there's a good reason we chose to use `Model.findByIdAndUpdate` instead of `Model.update` in the past. I'm OK with using `Model.update`, but I want to make sure that making the change cause problems 

@kevinwang2040 @Anand-Issac do you guys remember why we use `Model.findByIdAndUpdate` instead of `Model.update`?

> Also, note that if dateUpdated is set on the object passed to update/findByIdAndUpdate then the timestamp may fail to update (so ensure that this property isn't set).

@kevinwang2040 I'll leave this for you to test. I suspect that even if `dateUpdated` is set on the object passed into `Model.findByIdAndUpdate`/`Model.update`, the timestamp will overwrite it.

See docs for  [`Model.findByIdAndUpdate`](https://mongoosejs.com/docs/api.html#model_Model.findByIdAndUpdate) and [`Model.update`](https://mongoosejs.com/docs/api.html#model_Model.update).

You can pass `options` as the third param of `Model.findByIdAndUpdate` and `Model.update`. "If `options.timstamps != false`  and schema-level timestamps are enabled, skip timestamps for this update. "

Reply:
Re: timestamps being overwritten, this used to be the case and was a bug, but apparently it was patched: https://github.com/Automattic/mongoose/issues/4768#issuecomment-265287110

Nvm then